x-pack/plugins/endpoint/server/index_pattern.ts

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { SavedObjectsClientContract, Logger, LoggerFactory } from 'kibana/server';
+import { IndexPatternService } from '../../ingest_manager/server';
+
+const endpointPackageName = 'endpoint';
+
+export interface IndexPatternRetriever {
+  get(client: SavedObjectsClientContract, datasetPath: string): Promise<string>;
+}
+
+export class IngestIndexPatternRetriever implements IndexPatternRetriever {
+  private readonly log: Logger;
+  constructor(private readonly service: IndexPatternService, loggerFactory: LoggerFactory) {
+    this.log = loggerFactory.get('index-pattern-retriever');
+  }
+
+  async get(client: SavedObjectsClientContract, datasetPath: string, version?: string) {
+    const pattern = await this.service.get(client, endpointPackageName, datasetPath);
```

Comment:
So the call stack in Ingest is **route -> handler -> service**. The implementation of IndexPatternService is at the service layer, and the point I am trying to make is that some handlers may call more than one service. This appears to be a new service, so it may be fine. But when using an existing services e.g. https://github.com/elastic/kibana/blob/master/x-pack/plugins/ingest_manager/server/routes/agent/handlers.ts#L40 , which I will consuming for Agent Status information may need to account for or wrap already existing code in the handler. Such code may have to be at the handler level or code in the handler pushed down to the service level. This may mean that we may have to shift code around to avoid duplication and maintenance challenges.  @nchaulet @jfsiii @ruflin do you share this concern and the change implications for future changes using this approach?

Reply:
@nnamdifrankie I am not sure I understand the question. For the reasons you mentioned, the preference is to use the HTTP endpoint. We don't yet support services as contracts with other plugins, they are internal implementation details. However, we may need to export them or some other way to construct/execute the HTTP request.

If possible, I suggest Endpoint start with the HTTP request. If you are doing it in the browser, we have lots of hooks like https://github.com/elastic/kibana/blob/b38c98a730f4962910e29acf7d8847346ab67e6d/x-pack/plugins/ingest_manager/public/applications/ingest_manager/hooks/use_request/agents.ts#L21 which only need a little info and they'll execute & return the response.

Let me know if I misunderstood or if you need anything else.

Reply:
@jfsiii 

> I am not sure I understand the question. For the reasons you mentioned, the preference is to use the HTTP endpoint. We don't yet support services as contracts with other plugins, they are internal implementation details. However, we may need to export them or some other way to construct/execute the HTTP request.

I think you may have answered my question a bit. I wanted to know if ingest supports this kind of export at the service layer. My concern is that some handlers use more than one service and are rich, and exporting at the service layer may need to account for all the calls in the handler. 

> If possible, I suggest Endpoint start with the HTTP request. If you are doing it in the browser, we have lots of hooks like

We will not always do it from the browser. We will need to make server side calls to Ingest. Do we expose the hooks via the Plugin Dependencies Contract which in turns makes http call and invariably go through the right entry point (handler)?

Reply:
> > If possible, I suggest Endpoint start with the HTTP request. If you are doing it in the browser, we have lots of hooks like
> 
> We will not always do it from the browser. We will need to make server side calls to Ingest. Do we expose the hooks via the Plugin Dependencies Contract which in turns makes http call and invariably go through the right entry point (handler)?

@nnamdifrankie My understanding of the purpose of the kibana platform's plugin contracts is to avoid having to make direct HTTP calls into another plugin. I think John is suggesting to hold off on defining anymore ingest plugin contracts and use rest calls instead.

Reply:
@jonathan-buttner Thanks I will look into use the core http client, it may be a quicker resolution. But we still need the plugin dependency so that it guarantees that ingest is loaded and our http calls does not bomb out.